library/std/src/net/parser.rs

```diff
@@ -127,7 +129,16 @@ impl<'a> Parser<'a> {
                 }
             }
 
-            if digit_count == 0 { None } else { Some(result) }
+            if digit_count == 0 {
+                None
+            } else if !allow_zero_prefix
+                && has_leading_zero
+                && (result != T::ZERO || result == T::ZERO && digit_count > 1)
```

Comment:
I think this can be simplified as the following:

```suggestion
            } else if !allow_zero_prefix && has_leading_zero && digit_count > 1
```

With this, you also don't need to add `crate::cmp::PartialEq` bound to `ReadNumberHelper`